### DISCUSSION TITLE
Model keepSnapshots(true) not respected when manually assigning data to a model instance

### DISCUSSION THREAD

**Discussion by richmilns** [phalcon/cphalcon#15808] [Q&A] [answered]
I appreciate this is a bit of a niche issue, but I've run into what feels like a bug with keeping snapshots of model data when manually assigning data to a model via with the model `__construct()` method or `assign()` method.

I'm using Phalcon 4.1.2 with PHP 7.4.25 and have found that when you manually assign data to an instance of a `Model`, that the `keepSnapshots(true)` setting I have enabled is ignored and an exception is thrown:

```
Phalcon\Mvc\Model\Exception: The 'keepSnapshots' option must be enabled to track changes
Source: phalcon/Mvc/Model.zep on line 1674
```

Below is a simplified example to show how to replicate the problem I've run into.

I have a model called `Members` which has a `belongsTo()` relationship to another model called `Users`. 

My models look like this:

```php
namespace App;

class Members extends \Phalcon\Mvc\Model
{
    public $id;
    public $title;
    public $user_id;

    public function initialize()
    {
        $this->setSource('members');
        $this->keepSnapshots(true);
        $this->belongsTo(
            'user_id', 
            Users::class, 
            'id', 
             ['alias' => 'User']
        );
    }

    public function afterUpdate()
    {
         ray($this->getSnapshotData());
         ray($this->getChangedFields());
    }
}

class Users extends \Phalcon\Mvc\Model
{
    public $id;
    public $name;
    public $deleted;
    
    public function initialize()
    {
         $this->setSource('users');
    }
}

```

I am calling `getSnapshotData()` and `getChangedFields()` in my Members `afterUpdate()` method to see what changed and send an email notification in some cases.

I want to verify ownership of records before users can view or update a record.

An easy way to do this would be to join onto my Users model using my User ID to see if the record data is returned from the DB.

To do this, I could use Phalcon's query builder in my controller update action to get the record. I've skipped checks on whether the record exists to keep things more concise:

```php
$builder = $this->modelsManager
   ->createBuilder()
   ->columns('m.*')
   ->from(['m' => Members::class])
   ->andWhere('m.id = :id: AND m.user_id = :user_id:', [
   		'id' => $this->request->getQuery('id'),
		'user_id' => 1
   ])
   ->innerJoin(Users::class, 'u.id = m.user_id AND u.deleted = 0', 'u');

$target = $builder->getQuery()->execute()->getFirst();
$record = new Members($target->toArray());
var_dump($record); // data is correctly assigned to the model instance
$record->title = $this->request->getPost('title');
$record->update();
// exception gets thrown "The 'keepSnapshots' option must be enabled to track changes"
```

If I skip all my access checks and simply use `Members::findFirstById()` then no exception is thrown and the snapshot data is available `afterSave()`:

```php
$record = Members::findFirstById($this->request->getQuery('id'));
$record->title = $this->request->getPost('title');
$record->update();
// no exception thrown, snapshots are being kept and record updated
```

It would be awesome if it was possible to perform a custom DB query, manually assign the data back to the model and update as required whilst keeping snapshots. The only workaround I have is to do my access checks and then `findFirstById()` to get the record afterwards, which is a duplicate DB query and ideally should be avoidable.

Am I going about this the wrong way? Is this a bug? Thanks in advance for your thoughts.

**Comment by zsilbi** [comment, marked as answer]
This is unnecessary:
```php
$target = $builder->getQuery()->execute()->getFirst();
$record = new Members($target->toArray());
```

If your query is correct, `$target` should be a `Members` object which can be updated right away.

```php
$builder = $this->modelsManager
   ->createBuilder()
   ->columns('m.*')
   ->from(['m' => Members::class])
   ->andWhere('m.id = :id: AND m.user_id = :user_id:', [
   		'id' => $this->request->getQuery('id'),
		'user_id' => 1
   ])
   ->innerJoin(Users::class, 'u.id = m.user_id AND u.deleted = 0', 'u');

$target = $builder->getQuery()->execute()->getFirst();
$target->title = $this->request->getPost('title');
$target->update();
```

**Reply by richmilns** [reply]
You're right! Thanks for your help - the key for me though was to make sure that I don't specify any columns other than those belonging to the `Member` model in the example above, otherwise you don't get an instance of a `Phalcon\Mvc\Model` back, you get an instance of `Phalcon\Mvc\Model\Row`.

**Reply by quasipickle** [reply]
I don't understand why the unnecessary code would have triggered an exception.  That code should have caused `keepSnapshots()` to be set too.

**Reply by richmilns** [reply]
Exactly. I thought the same. I think there has to be a bug somewhere but I want to do a bit more testing.
